dialects/oracle/oracle.go

```diff
@@ -0,0 +1,223 @@
+	}
+	return false
+}
+
+func (s oracle) HasTable(tableName string) bool {
+	var count int
+	_, tableName = currentDatabaseAndTable(&s, tableName)
+	tableName = strings.ToUpper(tableName)
+	if err := s.db.QueryRow("select count(*) from user_tables where table_name = :1", tableName).Scan(&count); err == nil {
+		return count > 0
+	} 
+	return false
+}
+
+func (oracle) LastInsertIDReturningSuffix(tableName, columnName string) string {
+	return ""
+}
+
+func (oracle) LastInsertIDOutputInterstitial(tableName, columnName string, columns []string) string {
+	return ""
+}
+
+func (s oracle) ModifyColumn(tableName string, columnName string, typ string) error {
+	_, err := s.db.Exec(fmt.Sprintf("ALTER TABLE %v MODIFY %v %v", tableName, columnName, typ))
+	return err
+}
+
+func (s oracle) RemoveIndex(tableName string, indexName string) error {
+	_, err := s.db.Exec(fmt.Sprintf("DROP INDEX %v", indexName))
+	return err
+}
+
+func (oracle) SelectFromDummyTable() string {
+	return "FROM DUAL"
+}
+
+func (s *oracle) SetDB(db gorm.SQLCommon) {
+	s.db = db
+}
+
+func currentDatabaseAndTable(dialect gorm.Dialect, tableName string) (string, string) {
+	if strings.Contains(tableName, ".") {
+		splitStrings := strings.SplitN(tableName, ".", 2)
+		return splitStrings[0], splitStrings[1]
+	}
+	return dialect.CurrentDatabase(), tableName
+}
+
+func (s *oracle) DataTypeOf(field *gorm.StructField) string {
+	if _, found := field.TagSettingsGet("RESTRICT"); found {
+		field.TagSettingsDelete("RESTRICT")
+	}
+	var dataValue, sqlType, size, additionalType = gorm.ParseFieldStructForDialect(field, s)
+
+	if sqlType == "" {
+		switch dataValue.Kind() {
+		case reflect.Bool, reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Uint, reflect.Uint8,
+			reflect.Uint16, reflect.Uintptr, reflect.Int64, reflect.Uint32, reflect.Uint64,
+			reflect.Float32, reflect.Float64:
+			if s.fieldCanAutoIncrement(field) {
+				sqlType = "NUMBER GENERATED BY DEFAULT AS IDENTITY"
+			} else {
+				sqlType = "NUMBER"
+			}
+		case reflect.String:
+			if _, ok := field.TagSettingsGet("SIZE"); !ok {
+				size = 0 // if SIZE haven't been set, use `text` as the default type, as there are no performance different
+			}
+			switch {
+			case size > 0 && size < 4000:
+				sqlType = fmt.Sprintf("VARCHAR2(%d)", size)
+			case size == 0:
+				sqlType = "VARCHAR2 (4000)"
+			default:
+				sqlType = "CLOB"
+			}
+
+		case reflect.Struct:
+			if _, ok := dataValue.Interface().(time.Time); ok {
+				sqlType = "TIMESTAMP WITH TIME ZONE"
+			}
+		case reflect.Map:
+			if dataValue.Type().Name() == "Hstore" {
+				sqlType = "hstore"
+			}
+		default:
+			if gorm.IsByteArrayOrSlice(dataValue) {
+				sqlType = "VARCHAR2 (4000)"
+			}
+		}
+	}
+
+	if sqlType == "" {
+		panic(fmt.Sprintf("invalid sql type %s (%s) for oracle", dataValue.Type().Name(), dataValue.Kind().String()))
+	}
+
+	if strings.TrimSpace(additionalType) == "" {
+		return sqlType
+	}
+	if strings.EqualFold(sqlType, "json") {

```

Comment:
Wouldn't [panic have been called above](https://github.com/jimlambrt/gorm/blob/57dda4b080cb2d28c75b6b9c677ffb2fabad9e45/dialects/oracle/oracle.go#L180-L182) before getting to here?

Reply:
I don't think so... the sqlType is set.  The panic would happen if it was an empty string.